src/solc_0.6/EstateSale/EstateSaleWithAuth.sol

```diff
@@ -0,0 +1,254 @@
+    /// @param size size of the pack of Land to purchase
+    /// @param priceInSand price in SAND to purchase that Land
+    /// @param proof merkleProof for that particular Land
+    function buyLandWithSand(
+        /// TODO: review any other workaround for stack too deep
+        address[] calldata addresses,
+        uint256 x,
+        uint256 y,
+        uint256 size,
+        uint256 priceInSand,
+        bytes32 salt,
+        uint256[] calldata assetIds,
+        bytes32[] calldata proof,
+        bytes calldata referral,
+        bytes calldata signature
+    ) external {
+        _checkValidity2(addresses);
+        _checkValidity(addresses, x, y, size, priceInSand, salt, assetIds, proof, signature);
+        _handleFeeAndReferral(addresses[0], priceInSand, referral);
+        _mint(addresses, x, y, size, priceInSand, priceInSand);
+        _sendAssets(addresses[1], assetIds);
+    }
+
+    /// @notice Gets the expiry time for the current sale
+    /// @return The expiry time, as a unix epoch
+    function getExpiryTime() external view returns (uint256) {
+        return _expiryTime;
+    }
+
+    /// @notice Gets the Merkle root associated with the current sale
+    /// @return The Merkle root, as a bytes32 hash
+    function getMerkleRoot() external view returns (bytes32) {
+        return _merkleRoot;
+    }
+
+    /// @notice enable Admin to withdraw remaining assets from EstateSaleWithFee contract
+    /// @param to intended recipient of the asset tokens
+    /// @param assetIds the assetIds to be transferred
+    /// @param values the quantities of the assetIds to be transferred
+    function withdrawAssets(
+        address to,
+        uint256[] calldata assetIds,
+        uint256[] calldata values
+    ) external {
+        require(msg.sender == _admin, "NOT_AUTHORIZED");
+        // require(block.timestamp > _expiryTime, "SALE_NOT_OVER"); // removed to recover in case of misconfigured sales
+        _asset.safeBatchTransferFrom(address(this), to, assetIds, values, "");
+    }
+
+    function onERC1155Received(
+        address, /*operator*/
+        address, /*from*/
+        uint256, /*id*/
+        uint256, /*value*/
+        bytes calldata /*data*/
+    ) external pure returns (bytes4) {
+        return 0xf23a6e61;
+    }
+
+    function onERC1155BatchReceived(
+        address, /*operator*/
+        address, /*from*/
+        uint256[] calldata, /*ids*/
+        uint256[] calldata, /*values*/
+        bytes calldata /*data*/
+    ) external pure returns (bytes4) {
+        return 0xbc197c81;
+    }
+
+    function _sendAssets(address to, uint256[] memory assetIds) internal {
+        uint256[] memory values = new uint256[](assetIds.length);
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            values[i] = 1;
+        }
+        _asset.safeBatchTransferFrom(address(this), to, assetIds, values, "");
+    }
+
+    function _checkValidity2(address[] memory addresses) internal view {
```

Comment:
Not the greatest name. Rename or merge it if it's possible

Reply:
There was an issue with stack too deep, I tried to keep the original names, will see what comes to mind

Reply:
`_checkAuthValidity` for one
`_checkSaleValidity` maybe for the other one?

Reply:
Updated!